sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/MetricsPCollectionConsumerRegistry.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.data;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfo;
+import org.apache.beam.runners.core.metrics.ExecutionStateTracker;
+import org.apache.beam.runners.core.metrics.MetricsContainerImpl;
+import org.apache.beam.runners.core.metrics.MetricsContainerStepMapEnvironment;
+import org.apache.beam.runners.core.metrics.SimpleExecutionState;
+import org.apache.beam.runners.core.metrics.SimpleMonitoringInfoBuilder;
+import org.apache.beam.runners.core.metrics.SimpleStateRegistry;
+import org.apache.beam.sdk.fn.data.FnDataReceiver;
+import org.apache.beam.sdk.metrics.MetricsEnvironment;
+import org.apache.beam.sdk.util.WindowedValue;
+
+/**
+ * The {@code MetricsPCollectionConsumerRegistry} is used to maintain a collection of consuming
+ * FnDataReceiver for each pCollectionId. Registering with this class allows inserting an element
+ * count counter for every pCollection. The consumers are wrapped with extra logic to collect user
+ * metrics, execution time metrics and element count metrics.
+ */
+public class MetricsPCollectionConsumerRegistry extends PCollectionConsumerRegistry {
+
+  private Map<String, ElementCountFnDataReceiver> pCollectionIdsToWrappedConsumer;
+  private SimpleStateRegistry executionStates = new SimpleStateRegistry();
+
+  public MetricsPCollectionConsumerRegistry() {
```

Comment:
I made this class access its ExecutionStateTracker and MetricContainerStepMap from the thead local variable injected in MetricsBeamFnDataClient. This removed the params from the ctor :), but it now relies on the thread local variable.

I also put some of the logic in a base class PCollectionConsumerRegistry, as most tests use this class just to setup the consumers properly, and do not need metrics. Most of the unit tests use the base PCollectionConsumerRegistryclass, eliminating the need to setup the thead local variables in those tests.